src/main/java/io/smartcat/ranger/core/ValueProxy.java

```diff
@@ -0,0 +1,47 @@
+package io.smartcat.ranger.core;
+
+/**
+ * Proxy around value that can cache value and can reset cache.
+ *
+ * @param <T> Type this value would evaluate to.
+ */
+public class ValueProxy<T> extends Value<T> {
+
+    private Value<T> delegate;
+
+    /**
+     * Constructs proxy without delegate.
+     */
+    public ValueProxy() {
+        this(null);
+    }
+
+    /**
+     * Constructs proxy with specified <code>delegate</code>.
+     *
+     * @param delegate Value which will be evaluated and cached.
+     */
+    public ValueProxy(Value<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Sets value to this proxy.
+     *
+     * @param delegate Value which will be evaluated and cached.
+     */
+    public void setDelegate(Value<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        delegate.reset();
```

Comment:
Should we avoid NullPointerException by using an explicit check and throw a more appropriate exception?